include/rocksdb/data_structure.h

```diff
@@ -15,37 +15,183 @@
 
 namespace ROCKSDB_NAMESPACE {
 
-// This is a data structure specifically designed as a "Set" for a
-// pretty small scale of Enum structure. For now, it can support up
-// to 64 element, and it is expandable in the future.
-template <typename ENUM_TYPE, ENUM_TYPE MAX_VALUE>
+namespace detail {
+int CountTrailingZeroBitsForSmallEnumSet(uint64_t);
+}  // namespace detail
+
+// Represents a set of values of some enum type with a small number of
+// possible enumerators. For now, it supports enums where no enumerator
+// exceeds 63 when converted to int.
+template <typename ENUM_TYPE, ENUM_TYPE MAX_ENUMERATOR>
 class SmallEnumSet {
+ private:
+  using StateT = uint64_t;
+  static constexpr int kStateBits = sizeof(StateT) * 8;
+  static constexpr int kMaxMax = kStateBits - 1;
+  static constexpr int kMaxValue = static_cast<int>(MAX_ENUMERATOR);
+  static_assert(kMaxValue >= 0);
+  static_assert(kMaxValue <= kMaxMax);
+
  public:
+  // construct / create
   SmallEnumSet() : state_(0) {}
 
+  template <class... TRest>
+  /*implicit*/ constexpr SmallEnumSet(const ENUM_TYPE e, TRest... rest) {
+    *this = SmallEnumSet(rest...).With(e);
+  }
+
+  // Return the set that includes all valid values, assuming the enum
+  // is "dense" (includes all values converting to 0 through kMaxValue)
+  static constexpr SmallEnumSet All() {
+    StateT tmp = StateT{1} << kMaxValue;
+    return SmallEnumSet(RawStateMarker(), tmp | (tmp - 1));
+  }
+
+  // destruct
   ~SmallEnumSet() {}
 
-  // Return true if the input enum is included in the "Set" (i.e., changes the
-  // internal scalar state successfully), otherwise, it will return false.
-  bool Add(const ENUM_TYPE value) {
-    static_assert(MAX_VALUE <= 63, "Size currently limited to 64");
-    assert(value >= 0 && value <= MAX_VALUE);
-    uint64_t old_state = state_;
-    uint64_t tmp = 1;
-    state_ |= (tmp << value);
-    return old_state != state_;
+  // copy
+  SmallEnumSet(const SmallEnumSet& that) = default;
+  SmallEnumSet& operator=(const SmallEnumSet& that) = default;
+
+  // move
+  SmallEnumSet(SmallEnumSet&& other) noexcept = default;
+  SmallEnumSet& operator=(SmallEnumSet&& other) noexcept = default;
```

Comment:
To follow the rule of zero, we could consider removing all these and let the compiler generate them. (The same goes below for the iterator.)